### PR TITLE
feat: Create monthly amount table and write to it

### PR DESCRIPTION
**Pull request by johevemi**
# Description
Add monthly amounts table and write to it

## Pull-request quality

- [x] The title adheres to [this guide](https://github.com/Mech0z/GitHubGuidelines)
- [x] Tests are written and executed locally
- [x] Subsystem tests have been tested (by manually deploying to `dev_003`) [here](https://github.com/Energinet-DataHub/dh3-environments/actions/runs/9345462296)

### CODE REVIEW

File: source/databricks/calculation_engine/package/datamigration/migration_scripts/202405231400_Create_monthly_amounts_table.sql
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/datamigration/migration_scripts/202405231400_Create_monthly_amounts_table.sql:1]: We unit-test possible values for the other tables. So we should probably do it for this table as well.

Comment (johevemi) [reply source/databricks/calculation_engine/package/datamigration/migration_scripts/202405231400_Create_monthly_amounts_table.sql:1]: yeah, but does it look right? I don't want to write test for a table that is not correct

Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/datamigration/migration_scripts/202405231400_Create_monthly_amounts_table.sql:1]: Just checking my understanding: Does this table contain the results of calculation step 1-6 in the last table on [this page](https://energinet.atlassian.net/wiki/spaces/D3/pages/253886474/Engrosberegninger)?

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    calculation_execution_time_start TIMESTAMP NOT NULL,
+
+    -- 36 characters UUID
+    calculation_result_id STRING NOT NULL,
+
+    grid_area_code STRING NOT NULL,
+    energy_supplier_id STRING,
+    time TIMESTAMP NOT NULL,
+    amount DECIMAL(18, 6),
+    charge_owner_id STRING,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:16]: charge_owner_id, charge_type and charge_code can be NOT NULL. Do you agree @khatozen ?

File: source/databricks/calculation_engine/package/datamigration/migration_scripts/202405231400_Create_monthly_amounts_table.sql [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    calculation_execution_time_start TIMESTAMP NOT NULL,
+    
+    -- 36 characters UUID
+    calculation_result_id STRING NOT NULL,
+
+    grid_area_code STRING NOT NULL,
+    energy_supplier_id STRING,
+    time TIMESTAMP NOT NULL,
+    amount DECIMAL(18, 6),
+    charge_owner_id STRING,
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/datamigration/migration_scripts/202405231400_Create_monthly_amounts_table.sql:16]:
```suggestion
    energy_supplier_id STRING NOT NULL,
    time TIMESTAMP NOT NULL,
    amount DECIMAL(18, 6) NOT NULL,
    charge_owner_id STRING NOT NULL,
```

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    calculation_id STRING NOT NULL,
+    -- Enum
+    calculation_type STRING NOT NULL,
+    calculation_execution_time_start TIMESTAMP NOT NULL,
+
+    -- 36 characters UUID
+    calculation_result_id STRING NOT NULL,
+
+    grid_area_code STRING NOT NULL,
+    energy_supplier_id STRING,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:13]: I think this is should be NOT NULL, right @khatozen ?

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+)
+USING DELTA
+TBLPROPERTIES (
+    delta.deletedFileRetentionDuration = 'interval 30 days',
+    delta.constraints.calculation_id_chk = "LENGTH ( calculation_id ) = 36",
+    delta.constraints.calculation_type_chk = "calculation_type IN ( 'WholesaleFixing', 'FirstCorrectionSettlement', 'SecondCorrectionSettlement', 'ThirdCorrectionSettlement' )",
+    delta.constraints.calculation_result_id_chk = "LENGTH ( calculation_result_id ) = 36",
+    delta.constraints.grid_area_code_chk = "LENGTH ( grid_area_code ) = 3",
+    delta.constraints.energy_supplier_id_chk = "LENGTH ( energy_supplier_id ) = 13 OR LENGTH ( energy_supplier_id ) = 16",
+    delta.constraints.charge_owner_id_chk = "charge_owner_id IS NULL OR LENGTH ( charge_owner_id ) = 13 OR LENGTH ( charge_owner_id ) = 16",
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:28]: remove IS NULL OR?

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql (hunk 1 above)
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:16]: Not type and code as I read [this page](https://energinet.atlassian.net/wiki/spaces/D3/pages/253886474/Engrosberegninger). Step 1-6 in the last table.

Comment (JesperGomes) [reply source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:16]: This table only includes step 1-4 in that table. Not sure how you see that code and type can be null. AFAIK that is only relevant for total monthly sums (which is another table)

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:16]: You're right. Not step 5 and 6. I withdraw all my comments related to this mistake 🙈

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql (hunk 3 above)
Comment (johevemi) [reply source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:13]: change to not null

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql (hunk 4 above)
Comment (johevemi) [reply source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:28]: removed

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    -- 36 characters UUID
+    calculation_result_id STRING NOT NULL,
+
+    grid_area_code STRING NOT NULL,
+    energy_supplier_id STRING NOT NULL,
+    quantity_unit STRING NOT NULL,
+    time TIMESTAMP NOT NULL,
+    amount DECIMAL(18, 6),
+    is_tax BOOLEAN,
+    charge_code STRING,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:18]: NOT NULL

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    -- 36 characters UUID
+    calculation_result_id STRING NOT NULL,
+
+    grid_area_code STRING NOT NULL,
+    energy_supplier_id STRING NOT NULL,
+    quantity_unit STRING NOT NULL,
+    time TIMESTAMP NOT NULL,
+    amount DECIMAL(18, 6),
+    is_tax BOOLEAN,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:17]: NOT NULL

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql (hunk 5 above)
Comment (johevemi) [reply source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql:18]: wholesale results does not have NOT NULL on these two

Comment (johevemi) [reply source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql]: changed to NOT NULL

File: source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql (hunk 6 above)
Comment (johevemi) [reply source/databricks/calculation_engine/package/datamigration/current_state_scripts/tables/create_monthly_amounts.sql]: change to NOT NULL

File: source/databricks/calculation_engine/package/calculation/output/monthly_amounts_storage_model_factory.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        col(Colname.charge_code).alias(MonthlyAmountsColumnNames.charge_code),
+        col(Colname.charge_type).alias(MonthlyAmountsColumnNames.charge_type),
+        col(Colname.charge_owner).alias(MonthlyAmountsColumnNames.charge_owner_id),
+    )
+
+
+def _get_column_group_for_calculation_result_id() -> list[str]:
+    return [
+        Colname.calculation_id,
+        Colname.charge_owner,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/calculation/output/monthly_amounts_storage_model_factory.py:60]: what about charge_code and charge_type?

Comment (johevemi) [reply source/databricks/calculation_engine/package/calculation/output/monthly_amounts_storage_model_factory.py:62]: add both  and the missing tests for it

**Review by JesperGomes — DISMISSED** [review]

**Review by JesperGomes — APPROVED** [review]

**Pull request merged**